README.md
# STM Shuttle Tracker
Tracks STM "École" buses. [Try it out today!](https://stm.austinhuang.me)

Much thanks to [@felixinx](https://github.com/felixinx) for data assistance!

## Rebuild procedure
Please note that rehosting is only allowed for development purposes (e.g. improving this service).

* You need an environment variable named `STM_API_KEY`. The content should be the API key acquired from [STM Developer Portal](https://developpeurs.stm.info).
* Edit line 23 (time zone) of server.js accordingly.
* `npm i`, `npm start`, and you should be good to go!


changelog.txt
https://stm.austinhuang.me

Inquiries? Feedbacks? Concerns? im@austinhuang.me

=== C H A N G E L O G ===

Version 0.1.6, 2021/08/21
* Tracker
- Change license to GPLv3.

Version 0.1.5, 2020/04/10
* Schools
- Temporarily adding Académie Dunton and Greaves Adventist Academy

Version 0.1.4, 2020/04/09
* Schools
- Temporarily adding Pierre-Laporte

Version 0.1.3, 2020/03/31
* Tracker
- Refactored code for smooth operation

Version 0.1.2, 2020/03/28
* Tracker
- Complete revamp, stabilized vehicle tracking

Version 0.1.1, 2020/03/26
* Tracker
- Hotfix for 0.1.0 regarding formats and stuff
* Schools
- Permanently adding James Lyng and Mgr-Richard
- Temporarily adding Collège Français, thanks to a contributor

Version 0.1.0, 2020/03/25
* Tracker
- TripUpdates API implementation: Adding marker at termini to show upcoming departures
- Change of database structure
* Schools
- Temporarily adding Mgr-Richard
- Temporarily adding Collège Sainte-Anne (411E only), thanks to a contributor

--- 0.0.x ---

Version 0.0.8, 2020/03/21
* Tracker
- Every image is now proxied
- Added proper French for homepage only

Version 0.0.7, 2020/03/19
* Tracker
- Added proxy for OSM tiles, scripts, stylesheets, etc. Yes for privacy!
- Bug fix for database structure change

Version 0.0.6, 2020/03/18
* Tracker
- Added auto-refresh
- Change of database structure
- Some design changes with bilingual signages

Version 0.0.5, 2020/03/13
During coronavirus crisis, please stay safe.
* Tracker
- Information is now presented with pictograms and emojis rather than words

Version 0.0.4, 2020/03/11
* Tracker
- 45s interval => 30s interval due to STM approval
- Added departure time comparison
* Schools
- LCCHS/PACC now with full support (but with guessed schedule)
- Added Collège de Montréal

Version 0.0.3, 2020/03/10
* Tracker
- Added trip prediction (Marianopolis and André-Laurendeau only)
- Switched from humanize-duration to pretty-ms just to show negative time
- Finalized data interpreter
- OpenGraph implementation for Discord embed / Facebook link preview
* Schools
- Added LCCHS/PACC
- Fixed trip IDs for Marianopolis

Version 0.0.2, 2020/03/09
* Tracker
- This tracker became a *real* thing!
- Added map
- Added index.html
* Schools
- Mapped trip IDs for Marianopolis
- Added André-Laurendeau and Rosemont

Version 0.0.1, 2020/03/03
- This tracker became a thing...
- Experimented with raw data.


server.js
/*
    STM Shuttle Tracker
    Copyright (C) 2020-2021 Austin Huang <im@austinhuang.me> (https://austinhuang.me)

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

const express = require("express"),
  request = require("request"),
  DB = require("quick.db"),
  fs = require("fs"),
  time = "EDT", // Change to EST for non-daylight saving time!!!
  sample = fs.readFileSync("./sample.html", "utf8"),
  routes = new DB.table("routes"),
  routelist = require("./routes.json"),
  assets = require("./assets.json");
var app = express(),
  listener = app.listen(process.env.PORT, function() {
    console.log("Your app is listening on port " + listener.address().port);
  });

require("./update.js")(time);
setInterval(() => require("./update.js")(time), 30000);

app.get("/", (req, res) => {
  res.sendFile(__dirname + "/index.html");
});

app.get("/faq", (req, res) => {
  res.sendFile(__dirname + "/faq.html");
});

app.get("/:name.txt", (req, res) => {
  fs.access("./" + req.params.name + ".txt", fs.F_OK, err => {
    if (err) res.status(404).send("File not found");
    else res.sendFile(__dirname + "/" + req.params.name + ".txt");
  });
});

app.get("/assets/:name", (req, res) => {
  !assets[req.params.name]
    ? res.status(404).send("Not found in assets.json")
    : request(assets[req.params.name]).pipe(res);
});

app.get("/osm/:s/:z/:x/:y.png", (req, res) => {
  if (parseInt(req.params.z) < 10)
    return res.status(400).send("Zoom below 10 is rejected.");
  else if (
    parseInt(req.params.x) / 2 ** (parseInt(req.params.z) - 10) < 301 ||
    parseInt(req.params.x) / 2 ** (parseInt(req.params.z) - 10) > 303
  )
    return res
      .status(400)
      .send("Out of Montreal (x must be between 301 and 303 on zoom 10)");
  else if (
    parseInt(req.params.y) / 2 ** (parseInt(req.params.z) - 10) < 365 ||
    parseInt(req.params.y) / 2 ** (parseInt(req.params.z) - 10) > 367
  )
    return res
      .status(400)
      .send("Out of Montreal (y must be between 365 and 367 on zoom 10)");
  request(
    `https://${req.params.s}.tile.openstreetmap.org/${req.params.z}/${req.params.x}/${req.params.y}.png`,
    { headers: { "User-Agent": "I am caching for https://stm.austinhuang.me" } }
  ).pipe(res);
});

app.get("/:school", async (req, res) => {
  if (!routelist[req.params.school]) res.status(404).send("Invalid school.");
  else if (
    (time === "EDT" &&
      !routelist[req.params.school].period.find(
        p =>
          p[0] + 4 <= new Date().getUTCHours() &&
          p[1] + 4 >= new Date().getUTCHours()
      )) ||
    (time === "EST" &&
      !routelist[req.params.school].period.find(
        p =>
          p[0] + 5 <= new Date().getUTCHours() &&
          (p[1] + 5 === 24 ? 0 : p[1] + 5) >= new Date().getUTCHours()
      )) ||
    new Date().getDay() === 0 ||
    new Date().getDay() === 6
  )
    res.status(503).sendFile(__dirname + "/unavailable.html");
  else {
    let x = await routes.fetch(req.params.school),
      t = await DB.fetch("time." + req.params.school);
    if (!x)
      res
        .status(501)
        .send(
          "No bus data yet. This could be due to the school being relatively new to the system. Wait a couple of hours."
        );
    else
      res.send(
        sample
          .replace(
            "[BUSES]",
            [...x.loc.up, ...x.loc.down]
              .map(r => {
                let route = routelist[req.params.school].routes.find(
                  y => y.route === r.vehicle.trip.routeId
                );
                return (
                  "L.marker([" +
                  r.vehicle.position.latitude +
                  ", " +
                  r.vehicle.position.longitude +
                  '], {icon: bus}).addTo(mymap).bindPopup("<table style=\\"border-width:0px;\\"><tr><td align=\\"right\\">Bus #' +
                  r.id +
                  '</td><td align=\\"center\\">▼</td><td>🗒️ ' +
                  r.vehicle.trip.startTime +
                  '</td></tr><tr><td /><td align=\\"center\\">↓</td><td>' +
                  (r.vehicle.currentStatus === "STOPPED_AT"
                    ? ""
                    : "📡 " +
                      new Date(r.vehicle.timestamp * 1000).toLocaleString(
                        "en-US",
                        {
                          hour12: false,
                          timeZone: "America/Montreal",
                          hour: "2-digit",
                          minute: "2-digit",
                          second: "2-digit"
                        }
                      )) +
                  '</td></tr><tr><td align=\\"right\\">' +
                  route.stops[
                    (x.tr.up.indexOf(r.vehicle.trip.tripId) > -1 ? "u" : "d") +
                      r.vehicle.currentStopSequence
                  ] +
                  " (" +
                  r.vehicle.currentStopSequence +
                  ')</td><td align=\\"center\\">' +
                  (r.vehicle.currentStatus === "STOPPED_AT"
                    ? "⬤</td><td>📡 " +
                      new Date(r.vehicle.timestamp * 1000).toLocaleString(
                        "en-US",
                        {
                          hour12: false,
                          timeZone: "America/Montreal",
                          hour: "2-digit",
                          minute: "2-digit",
                          second: "2-digit"
                        }
                      ) +
                      "</td>"
                    : "◯</td><td>" +
                      (x.tu.up.find(
                        n => n.tripUpdate.trip.tripId === r.vehicle.trip.tripId
                      ) && route.up === "u" + r.vehicle.currentStopSequence
                        ? "🔮 [" +
                          new Date(
                            x.tu.up.find(
                              n =>
                                n.tripUpdate.trip.tripId ===
                                r.vehicle.trip.tripId
                            ) &&
                              x.tu.up.find(
                                n =>
                                  n.tripUpdate.trip.tripId ===
                                  r.vehicle.trip.tripId
                              ).tripUpdate.stopTimeUpdate[
                                parseInt(r.vehicle.currentStopSequence) - 1
                              ].arrival.time
                          )
                            .toLocaleString("en-US", {
                              timeZone: "America/Montreal"
                            })
                            .split(",")[0] +
                          "]"
                        : x.loc.down.find(
                            n => n.vehicle.trip.tripId === r.vehicle.trip.tripId
                          ) &&
                          route.down === "d" + r.vehicle.currentStopSequence
                        ? "🔮 [" +
                          new Date(
                            x.tu.down
                              .find(
                                n =>
                                  n.tripUpdate.trip.tripId ===
                                  r.vehicle.trip.tripId
                              )
                              .tripUpdate.stopTimeUpdate.find(
                                s =>
                                  s.stopSequence ===
                                  r.vehicle.currentStopSequence
                              ).arrival.time * 1000
                          ).toLocaleString("en-US", {
                            hour12: false,
                            timeZone: "America/Montreal",
                            hour: "2-digit",
                            minute: "2-digit",
                            second: "2-digit"
                          }) +
                          "]"
                        : "")) +
                  "</td>" +
                  (x.tr.up.indexOf(r.vehicle.trip.tripId) > -1 &&
                  route.up !== "u" + r.vehicle.currentStopSequence // up, not last stop?
                    ? '</tr><tr><td /><td align=\\"center\\">↓</td><td /></tr><tr><td align=\\"right\\">' +
                      route.stops[route.up] +
                      '</td><td align=\\"center\\">◯</td><td>' +
                      (x.tu.up.find(
                        n => n.tripUpdate.trip.tripId === r.vehicle.trip.tripId
                      ) &&
                      x.tu.up
                        .find(
                          n =>
                            n.tripUpdate.trip.tripId === r.vehicle.trip.tripId
                        )
                        .tripUpdate.stopTimeUpdate.find(
                          s =>
                            s.stopSequence === parseInt(route.up.substring(1))
                        )
                        ? "🔮 [" +
                          new Date(
                            x.tu.up
                              .find(
                                n =>
                                  n.tripUpdate.trip.tripId ===
                                  r.vehicle.trip.tripId
                              )
                              .tripUpdate.stopTimeUpdate.find(
                                s =>
                                  s.stopSequence ===
                                  parseInt(route.up.substring(1))
                              ).arrival.time * 1000
                          )
                            .toLocaleString("en-US", {
                              hour12: false,
                              timeZone: "America/Montreal",
                              hour: "2-digit",
                              minute: "2-digit",
                              second: "2-digit"
                            })
                            .split(",")[0] +
                          "]"
                        : "") +
                      "</td></tr>"
                    : x.tr.down.indexOf(r.vehicle.trip.tripId) > -1 &&
                      route.down !== "d" + r.vehicle.currentStopSequence // down, not last stop?
                    ? '</tr><tr><td /><td align=\\"center\\">↓</td><td /></tr><tr><td align=\\"right\\">' +
                      route.stops[route.down] +
                      '</td><td align=\\"center\\">◯</td><td>' +
                      (x.tu.down.find(
                        n => n.tripUpdate.trip.tripId === r.vehicle.trip.tripId
                      ) &&
                      x.tu.down
                        .find(
                          n =>
                            n.tripUpdate.trip.tripId === r.vehicle.trip.tripId
                        )
                        .tripUpdate.stopTimeUpdate.find(s => {
                          return (
                            s.stopSequence === parseInt(route.down.substring(1))
                          );
                        })
                        ? "🔮 [" +
                          new Date(
                            x.tu.down
                              .find(
                                n =>
                                  n.tripUpdate.trip.tripId ===
                                  r.vehicle.trip.tripId
                              )
                              .tripUpdate.stopTimeUpdate.find(
                                s =>
                                  s.stopSequence ===
                                  parseInt(route.down.substring(1))
                              ).arrival.time * 1000
                          ).toLocaleString("en-US", {
                            hour12: false,
                            timeZone: "America/Montreal",
                            hour: "2-digit",
                            minute: "2-digit",
                            second: "2-digit"
                          }) +
                          "]"
                        : "")
                    : "</tr>") +
                  '</table>");'
                );
              })
              .join("\n")
          )
          .replace("[STOPS]", () => {
            let upFroms = routelist[req.params.school].routes
              .map(route => {
                if (route.upFromLoc)
                  return (
                    "L.marker([" +
                    route.upFromLoc +
                    ']).addTo(mymap).bindPopup("<table style=\\"border-width:0px;\\"><tr><td align=\\"right\\">' +
                    route.route +
                    ' 🚍</td><td align=\\"center\\">→</td><td>' +
                    route.stops[route.up] +
                    "</td></tr><tr>" +
                    (x.tu.up.length === 0 ||
                    x.tu.up.filter(
                      a =>
                        a.tripUpdate.stopTimeUpdate.length ===
                        parseInt(route.up.substring(1))
                    ).length === 0
                      ? '<td align=\\"right\\">Pas de bus</td><td /><td>No buses</td>'
                      : x.tu.up
                          .filter(
                            a =>
                              a.tripUpdate.stopTimeUpdate.length ===
                              parseInt(route.up.substring(1))
                          )
                          .sort(
                            (a, b) =>
                              a.tripUpdate.stopTimeUpdate[0].departure.time -
                              b.tripUpdate.stopTimeUpdate[0].departure.time
                          )
                          .map(
                            t =>
                              '<td align=\\"right\\">' +
                              new Date(
                                t.tripUpdate.stopTimeUpdate[0].departure.time *
                                  1000
                              ).toLocaleString("en-US", {
                                hour12: false,
                                timeZone: "America/Montreal",
                                hour: "2-digit",
                                minute: "2-digit",
                                second: "2-digit"
                              }) +
                              " (🗒️ " +
                              t.tripUpdate.trip.startTime +
                              ")</td><td /><td>" +
                              (t.tripUpdate.stopTimeUpdate.find(
                                s =>
                                  s.stopSequence ===
                                  parseInt(route.down.substring(1))
                              ).departure
                                ? new Date(
                                    t.tripUpdate.stopTimeUpdate.find(
                                      s =>
                                        s.stopSequence ===
                                        parseInt(route.down.substring(1))
                                    ).arrival.time * 1000
                                  ).toLocaleString("en-US", {
                                    hour12: false,
                                    timeZone: "America/Montreal",
                                    hour: "2-digit",
                                    minute: "2-digit",
                                    second: "2-digit"
                                  })
                                : "???") +
                              "</td>"
                          )
                          .join("</tr><tr>")) +
                    '</tr></table>");'
                  );
              })
              .join("\n");
            if (routelist[req.params.school].downFromLoc)
              upFroms +=
                "\nL.marker([" +
                routelist[req.params.school].downFromLoc +
                ']).addTo(mymap).bindPopup("<table style=\\"border-width:0px;\\">' +
                routelist[req.params.school].routes
                  .map(
                    route =>
                      '<tr><td align=\\"right\\">' +
                      route.route +
                      ' 🚍</td><td align=\\"center\\">→</td><td>' +
                      route.stops[route.down] +
                      "</td></tr><tr>" +
                      (x.tu.down.length === 0 ||
                      x.tu.down.filter(
                        a =>
                          a.tripUpdate.stopTimeUpdate.length ===
                          parseInt(route.down.substring(1))
                      ).length === 0
                        ? '<td align=\\"right\\">Pas de bus</td><td /><td>No buses</td>'
                        : x.tu.down
                            .filter(
                              a =>
                                a.tripUpdate.stopTimeUpdate.length ===
                                parseInt(route.down.substring(1))
                            )
                            .sort(
                              (a, b) =>
                                a.tripUpdate.stopTimeUpdate[0].departure.time -
                                b.tripUpdate.stopTimeUpdate[0].departure.time
                            )
                            .map(
                              t =>
                                '<td align=\\"right\\">' +
                                new Date(
                                  t.tripUpdate.stopTimeUpdate[0].departure
                                    .time * 1000
                                ).toLocaleString("en-US", {
                                  hour12: false,
                                  timeZone: "America/Montreal",
                                  hour: "2-digit",
                                  minute: "2-digit",
                                  second: "2-digit"
                                }) +
                                " (🗒️ " +
                                t.tripUpdate.trip.startTime +
                                ")</td><td /><td>" +
                                (t.tripUpdate.stopTimeUpdate.find(
                                  s =>
                                    s.stopSequence ===
                                    parseInt(route.down.substring(1))
                                )
                                  ? new Date(
                                      t.tripUpdate.stopTimeUpdate.find(
                                        s =>
                                          s.stopSequence ===
                                          parseInt(route.down.substring(1))
                                      ).departure.time * 1000
                                    ).toLocaleString("en-US", {
                                      hour12: false,
                                      timeZone: "America/Montreal",
                                      hour: "2-digit",
                                      minute: "2-digit",
                                      second: "2-digit"
                                    })
                                  : "???") +
                                "</td>"
                            )
                            .join("</tr><tr>")) +
                      "</tr>"
                  )
                  .join("") +
                '</table>");';
            return upFroms;
          })
          .replace(
            "[TIME]",
            "<br><i>" +
              new Date(t).toLocaleString("en-US", {
                timeZone: "America/Montreal"
              }) +
              ".</i>"
          )
          .replace(/\[SCHOOL\]/g, routelist[req.params.school].name)
          .replace(/\[school\]/g, req.params.school)
          .replace("[REFRESH]", req.query.refresh === "true" ? "true" : "false")
          .replace(
            "[METAR]",
            req.query.refresh === "true"
              ? '<meta http-equiv="refresh" content="30" >'
              : ""
          )
          .replace("[CENTER]", routelist[req.params.school].center)
      );
  }
});


update.js
/*
    STM Shuttle Tracker
    Copyright (C) 2020-2021 Austin Huang <im@austinhuang.me> (https://austinhuang.me)

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

const request = require("request"),
  GtfsRealtimeBindings = require("gtfs-realtime-bindings").transit_realtime,
  DB = require("quick.db"),
  fs = require("fs"),
  routes = new DB.table("routes"),
  routelist = require("./routes.json"),
  realstops = require("./stops.json"),
  stopfile = fs.createWriteStream("./stops.txt", { flags: "a" }),
  gtfsstops = fs.createReadStream("./stops.txt");

// function streamToString(stream) {
//   const chunks = [];
//   return new Promise((resolve, reject) => {
//     stream.on("data", chunk => chunks.push(chunk));
//     stream.on("error", reject);
//     stream.on("end", () => resolve(Buffer.concat(chunks).toString("utf8")));
//   });
// }

module.exports = time => {
  if (
    (time === "EDT" &&
      new Date().getUTCHours() >= 10 &&
      new Date().getUTCHours() <= 23) ||
    (time === "EST" &&
      new Date().getUTCHours() >= 11 &&
      new Date().getDay() !== 0 &&
      new Date().getDay() !== 6)
  )
    request(
      "https://api.stm.info/pub/od/gtfs-rt/ic/v2/tripUpdates",
      {
        method: "GET",
        headers: { apikey: process.env.STM_API_KEY },
        encoding: null
      },
      (e, r, b) => {
        if (e || r.statusCode !== 200) return console.error(e);
        let updt = GtfsRealtimeBindings.FeedMessage.decode(b);
        request(
          "https://api.stm.info/pub/od/gtfs-rt/ic/v2/vehiclePositions",
          {
            method: "GET",
            headers: { apikey: process.env.STM_API_KEY },
            encoding: null
          },
          (e, r, b2) => {
            if (e || r.statusCode !== 200) return console.error(e);
            let feed = GtfsRealtimeBindings.FeedMessage.decode(b2);
            Object.keys(routelist).map(async s => {
              let d = await routes.fetch(s);
              if (!d) d = { tu: {}, loc: {}, tr: {} };
              let ups = updt.entity.filter(
                  u =>
                    (routelist[s].routes.find(l => {
                      return (
                        l.upFromId === u.tripUpdate.stopTimeUpdate[0].stopId &&
                        l.upToId ===
                          u.tripUpdate.stopTimeUpdate[
                            u.tripUpdate.stopTimeUpdate.length - 1
                          ].stopId &&
                        l.route === u.tripUpdate.trip.routeId
                      );
                    }) ||
                      (d.tr.up && d.tr.up.indexOf(u.tripUpdate.trip.tripId) > -1)) &&
                    u.tripUpdate.stopTimeUpdate[0].departure &&
                    u.tripUpdate.stopTimeUpdate[0].departure.time * 1000 +
                      60000 >
                      Date.now()
                ),
                downs = updt.entity.filter(
                  u =>
                    (routelist[s].routes.find(l => {
                      return (
                        l.downFromId ===
                          u.tripUpdate.stopTimeUpdate[0].stopId &&
                        l.downToId ===
                          u.tripUpdate.stopTimeUpdate[
                            u.tripUpdate.stopTimeUpdate.length - 1
                          ].stopId &&
                        l.route === u.tripUpdate.trip.routeId
                      );
                    }) ||
                    (d.tr.down && d.tr.down.indexOf(u.tripUpdate.trip.tripId) > -1)) &&
                    u.tripUpdate.stopTimeUpdate[0].departure &&
                    u.tripUpdate.stopTimeUpdate[0].departure.time * 1000 +
                      60000 >
                      Date.now()
                );
              d.tu.up = ups;
              d.tu.down = downs;
              d.tr = !d.tr ? {} : d.tr;
              d.tr.up =
                (!d.tr.up || d.tr.up.length === 0)
                  ? ups.map(b => b.tripUpdate.trip.tripId)
                  : [
                      ...new Set([
                        ...d.tr.up,
                        ...ups.map(b => b.tripUpdate.trip.tripId)
                      ])
                    ];
              d.tr.down =
                (!d.tr.down || d.tr.down.length === 0)
                  ? downs.map(b => b.tripUpdate.trip.tripId)
                  : [
                      ...new Set([
                        ...d.tr.down,
                        ...downs.map(b => b.tripUpdate.trip.tripId)
                      ])
                    ];
              let upBuses = feed.entity.filter(
                  f => d.tr.up.indexOf(f.vehicle.trip.tripId) > -1
                ),
                downBuses = feed.entity.filter(f => {
                  return d.tr.down.indexOf(f.vehicle.trip.tripId) > -1;
                });
              DB.set("time." + s, Date.now());
              d.loc.up = upBuses;
              d.loc.down = downBuses;
              routes.set(s, d);
            });

            /*updt.entity
              .filter(
                u =>
                  u.tripUpdate.trip.routeId.endsWith("E") ||
                  u.tripUpdate.trip.routeId.endsWith("I")
              )
              .map(t => {
                t.tripUpdate.stopTimeUpdate.map(s => {
                  if (realstops.indexOf(s.stopId) === -1)
                    streamToString(gtfsstops)
                      .then(gtfsstop => {
                        if (
                          gtfsstop.indexOf(s.stopId + "," + s.stopId + ",") ===
                          -1
                        )
                          stopfile.write(
                            s.stopId +
                              "," +
                              s.stopId +
                              "," +
                              t.tripUpdate.trip.routeId +
                              " trip " +
                              t.tripUpdate.trip.tripId +
                              " route " +
                              t.tripUpdate.stopTimeUpdate[0].stopId +
                              " => " +
                              t.tripUpdate.stopTimeUpdate[
                                t.tripUpdate.stopTimeUpdate.length - 1
                              ].stopId +
                              " depart " +
                              t.tripUpdate.trip.startDate +
                              " " +
                              t.tripUpdate.trip.startTime.replace(/:00$/g, "") +
                              ",,,https://github.com/austinhuang0131/stm-shuttle-tracker/wiki/Fake-STM-Stops,0,,0\n",
                            "utf8",
                            console.error
                          );
                      })
                      .catch(console.error);
                });
              });*/
          }
        );
      }
    );
};
